### PR TITLE
Fix to fail e2e tests gracefully if setup is not ideal

### PR DESCRIPTION
**Pull request by ashahi1**
Fixes issue #1463 and #1398

1. Added a fix to ensure E2E tests will run even with one datastore - #1463
2. Added the fix to verify the minimum required environment variables are properly initialized before running e2e test else fail the test run.
3. Made changes to inputparams.GetTestConfig() so we have only one instance of config and not each E2E test creating a new instance.

Tested locally: All tests passed other than swarm tests as there was no swarm configured
```
=> Running target test-e2e-runalways
OK: 6 passed
--- PASS: Test (120.40s)
PASS

=> Running target test-e2e-runonce
--- FAIL: Test (601.62s)
FAIL
OOPS: 23 passed, 3 skipped, 1 FAILED, 2 MISSED
exit status 1
FAIL	github.com/vmware/docker-volume-vsphere/tests/e2e	601.626s
make: *** [test-e2e-runonce] Error 1
```

### CODE REVIEW

File: tests/e2e/swarm_test.go [hunk 1]
```diff
@@ -46,8 +46,11 @@ type SwarmTestSuite struct {
 func (s *SwarmTestSuite) SetUpSuite(c *C) {
 	s.esxName = inputparams.GetEsxIP()
 	s.master = inputparams.GetSwarmManager1()
+	c.Assert(s.master, Not(IsNil), Commentf("Master node is nil"))
```
Comment (shuklanirdesh82) [inline tests/e2e/swarm_test.go]: nit: Please improve the error message like `Master node is not set`

File: tests/utils/inputparams/testparams.go [hunk 2]
```diff
@@ -96,52 +101,68 @@ func GetEndPoint2() string {
...
 // GetTestConfig - returns the configuration of IPs for the
 // ESX host, docker hosts and the datastores on the host
-func GetTestConfig() *TestConfig {
-	var config *TestConfig
-
+func getInstance() *TestConfig {
+	noVMName := "no such VM"
 	config = new(TestConfig)
 	config.EsxHost = os.Getenv("ESX")
+	if config.EsxHost == "" {
```
Comment (shuklanirdesh82) [inline tests/utils/inputparams/testparams.go]: as discussed offline .. should be replaced by `GetEsxIP()`

File: tests/e2e/swarm_test.go (hunk 1 above)
Comment (lipingxue) [inline tests/e2e/swarm_test.go]: +1

File: tests/e2e/swarm_test.go [hunk 3]
```diff
@@ -46,8 +46,11 @@ type SwarmTestSuite struct {
...
 	s.worker1 = inputparams.GetSwarmWorker1()
+	c.Assert(s.worker1, Not(IsNil), Commentf("Worker1 node is nil"))
```
Comment (lipingxue) [inline tests/e2e/swarm_test.go]: Same as above. "Worker1 node is not set".

File: tests/utils/inputparams/testparams.go [hunk 4]
```diff
@@ -96,52 +101,68 @@ func GetEndPoint2() string {
...
+		log.Fatal("ESX host not found. Stopping the test run.")
+	}
 	config.DockerHosts = append(config.DockerHosts, os.Getenv("VM1"))
 	config.DockerHosts = append(config.DockerHosts, os.Getenv("VM2"))
+	if config.DockerHosts[0] == "" && config.DockerHosts[1] == "" {
+		log.Fatal("No docker hosts found. Atleast one host is needed to run tests.")
+	}
 	config.DockerHostNames = append(config.DockerHostNames, esx.RetrieveVMNameFromIP(config.DockerHosts[0]))
 	config.DockerHostNames = append(config.DockerHostNames, esx.RetrieveVMNameFromIP(config.DockerHosts[1]))
+	if config.DockerHostNames[0] == noVMName && config.DockerHostNames[1] == noVMName {
```
Comment (lipingxue) [inline tests/utils/inputparams/testparams.go]: So do we allow the test to run if only one VM is set?

Comment (ashahi1) [inline tests/utils/inputparams/testparams.go]: Yes, we will allow it. It will be taken care as part of PR #1515

File: tests/e2e/swarm_test.go (hunk 3 above)
Comment (ashahi1) [inline tests/e2e/swarm_test.go]: Done.

File: tests/utils/inputparams/testparams.go [hunk 5]
```diff
@@ -96,52 +101,65 @@ func GetEndPoint2() string {
...
+		log.Fatalf("No names found for docker hosts - %s , %s ", config.DockerHosts[0], config.DockerHosts[1])
+	}
 	config.Datastores = esx.GetDatastoreList()
-
-	if config.DockerHostNames[0] == "" || config.DockerHostNames[1] == "" ||
-		len(config.Datastores) <= 1 {
-		return nil
+	if len(config.Datastores) < 1 {
+		log.Fatalf("No datastores found. Atleast one datastore is needed to run tests.")
```
Comment (shaominchen) [inline tests/utils/inputparams/testparams.go]: Nit: Atleast => At least

File: tests/utils/inputparams/testparams.go (hunk 4 above)
Comment (shaominchen) [inline tests/utils/inputparams/testparams.go]: If we allow one VM setup, we should just check one host name here.

File: tests/utils/inputparams/testparams.go [hunk 6]
```diff
@@ -153,3 +171,15 @@ func GetRandomNumber() string {
 	bytes := min + rand.Intn(max)
 	return strconv.Itoa(int(bytes))
 }
+
+// GetTestConfig - Creates one instance of TestConfig
+func GetTestConfig() *TestConfig {
+	if config == nil {
```
Comment (shaominchen) [inline tests/utils/inputparams/testparams.go]: An Idiomatic Singleton Approach in Go is using sync.Once:

```go
import (
    "sync"
)

type singleton struct {
}

var instance *singleton
var once sync.Once

func GetInstance() *singleton {
    once.Do(func() {
        instance = &singleton{}
    })
    return instance
}
```

File: tests/utils/inputparams/testparams.go [hunk 7]
```diff
@@ -96,52 +101,65 @@ func GetEndPoint2() string {
 
 // GetSwarmManager1 returns swarm manager node IP from the configured swarm cluster
 func GetSwarmManager1() string {
-	return os.Getenv("MANAGER1")
+	manager1 := os.Getenv("MANAGER1")
+	if manager1 == "" {
+		log.Printf("Manager node not found. Need this to run swarm test.")
```
Comment (ashahi1) [inline tests/utils/inputparams/testparams.go]: @shaominchen I have removed the log.Printf from here and keeping the assert in the swarm test. I do not think we should change it to log.Fatal as in that case, whole test run will stop. And keeping assert in swarm test will only stop the swarm test and not the whole test run.

File: tests/utils/inputparams/testparams.go [hunk 8]
```diff
@@ -96,52 +101,65 @@ func GetEndPoint2() string {
...
-
+func getInstance() *TestConfig {
+	noVMName := "no such VM"
 	config = new(TestConfig)
-	config.EsxHost = os.Getenv("ESX")
+	config.EsxHost = GetEsxIP()
 	config.DockerHosts = append(config.DockerHosts, os.Getenv("VM1"))
 	config.DockerHosts = append(config.DockerHosts, os.Getenv("VM2"))
+	if config.DockerHosts[0] == "" && config.DockerHosts[1] == "" {
+		log.Fatal("No docker hosts found. Atleast one host is needed to run tests.")
```
Comment (ashahi1) [inline tests/utils/inputparams/testparams.go]: Done

File: tests/utils/inputparams/testparams.go (hunk 5 above)
Comment (ashahi1) [inline tests/utils/inputparams/testparams.go]: Done

File: tests/utils/inputparams/testparams.go (hunk 7 above)
Comment (shaominchen) [inline tests/utils/inputparams/testparams.go]: That's not the case. After addressing #1515, swarm test and other complicated tests will be moved out of the basic precheckin test suite. So it doesn't matter if we use log.Fatal() or not here. When swarm test and other complicated tests are REQUIRED to run - for example, in CI - we still want to fail the whole test if these environment variables are not available.

File: tests/utils/inputparams/testparams.go (hunk 6 above)
Comment (ashahi1) [inline tests/utils/inputparams/testparams.go]: @shaominchen It was one of the approaches mentioned here: http://marcio.io/2015/07/singleton-pattern-in-go/
But looks like the preferred approach is to use sync.once. 
Will make this change.

File: tests/utils/inputparams/testparams.go [hunk 9]
```diff
@@ -96,52 +101,56 @@ func GetEndPoint2() string {
 
 // GetSwarmManager1 returns swarm manager node IP from the configured swarm cluster
 func GetSwarmManager1() string {
-	return os.Getenv("MANAGER1")
+	manager1 := os.Getenv("MANAGER1")
```
Comment (shaominchen) [inline tests/utils/inputparams/testparams.go]: let's revert this change - no need to introduce a local variable here. Same for below.

File: tests/utils/inputparams/testparams.go [hunk 10]
```diff
@@ -96,52 +101,56 @@ func GetEndPoint2() string {
...
-	config.EsxHost = os.Getenv("ESX")
+	config.EsxHost = GetEsxIP()
 	config.DockerHosts = append(config.DockerHosts, os.Getenv("VM1"))
 	config.DockerHosts = append(config.DockerHosts, os.Getenv("VM2"))
+	if config.DockerHosts[0] == "" && config.DockerHosts[1] == "" {
+		log.Fatal("No docker hosts found. At least one host is needed to run tests.")
+	}
 	config.DockerHostNames = append(config.DockerHostNames, esx.RetrieveVMNameFromIP(config.DockerHosts[0]))
 	config.DockerHostNames = append(config.DockerHostNames, esx.RetrieveVMNameFromIP(config.DockerHosts[1]))
+	if config.DockerHostNames[0] == noVMName && config.DockerHostNames[1] == noVMName {
```
Comment (shaominchen) [inline tests/utils/inputparams/testparams.go]: The AND logic here should be OR, since we want to check both. Please fix the logs message below appropriately.

File: tests/utils/inputparams/testparams.go (hunk 9 above)
Comment (ashahi1) [inline tests/utils/inputparams/testparams.go]: Done

File: tests/utils/inputparams/testparams.go (hunk 10 above)
Comment (ashahi1) [inline tests/utils/inputparams/testparams.go]: Done
